alf/environments/metadrive/geometry.py

```diff
@@ -0,0 +1,238 @@
+        of ``segment_resolution``.
+
+        This means that the length of each polyline will be the product of
+        ``segment_resolution`` and ``polyline_size`` unless it is the last
+        polyline. The last polyline is usually longer or smaller, depending on
+        the length of curve described by the lane.
+
+        Args:
+
+            lane: a MetaDrive lane as the reference for the target curve.
+            lateral: describes the lateral offset (ratio) of the target curve
+                with respect to the lane. It should be within [-0.5, 0.5]. For
+                example, if it is set to -0.5, it means that the target curve is
+                the right side boundary of the lane. Similarly, 0.5 is for the
+                left side boungdary, and 0.0 is for the cetner line of the lane.
+            category: an integer representing the category of the curve. For
+                example, whether it is a broken line or a solid line. Not all
+                polylines need to have a category, but it is necessary here
+                since the polyline represents an object from the map.
+            segment_resolution: together with ``polyline_size`` it describes how
+                to divided the target curve into polylines. See method docstring
+                for details. Unit in meters.
+            polyline_size: together with ``segment_resolution`` it describes how
+                to divided the target curve into polylines. See method docstring
+                for details. Unit in meters.
+
+        """
+        polyline_length = segment_resolution * polyline_size
+
+        # Trying to figure out how many polylines we can get from the target
+        # curve, whose length is ``lane.length``. In the usual case when the
+        # length is not perfectly divisible by the polyline's length, we will
+        # have to make it up by making the last polyline longer or shorter (in
+        # the shorter case, number of polylines is increased by 1).
+        num_polylines = int(lane.length / polyline_length)
+        if lane.length % polyline_length > 0.5:
+            num_polylines += 1
+
+        result = Polyline(
+            point=np.zeros((num_polylines, polyline_size + 1, 2),
+                           dtype=np.float32),
+            category=np.full((num_polylines, ), category, dtype=np.int32))
+
+        # Here s is the distance of the sampled point along the curve, from the
+        # starting point of the curve.
+        s = 0.0
+        sample_point = lane.position(s, lateral * lane.width_at(s))
+        for i in range(num_polylines):
+            result.point[i, 0] = sample_point
+
+            # Normally it samples a point every ``segment_resolution`` meters.
+            # However for the last segment since it does not have the standard
+            # length, we will have to improvise.
+            if i == num_polylines - 1:
+                seg_len = (lane.length - s) / polyline_size
```

Comment:
The suggested (bundle adjustment) is actually the alternative I was thinking about when implementing this. I ended up choosing the current approach because the following pros and cons:

1. Better readability and slightly simpler logic
2. Most polylines will have the standard length
3. (Cons) the polylines that are much shorter/longer becomes minority and may invite bias

I doubt if there can be noticeable difference speaking of the result. Let me run an experiment of the alternative to see.

Reply:
Point 3 is what I worry about. There can be problem if there is some critical event happening at that location. However, its effect is hard to quantify because of its rarity. 
I don't feel point 1 is a problem. In fact, there is no more if/else in the for loop for handling this, which is good.

Reply:
Point taken. I will update this PR after the experiment, just to make sure it still works as expected.

Reply:
The experiment with the alternative approach turns out to be slightly different in terms of how soon it converges. I'll check whether there are any issues. Visualization looks reasonable though.

Reply:
New experiment verified that there is no noticeable disadvantages or advantages, but I agree that it might be better. I created a separate PR for this change: #1186